django/db/models/sql/compiler.py

```diff
@@ -1787,7 +1815,10 @@ def as_sql(self):
                 result.append("VALUES (%s)" % ", ".join(placeholder_rows[0]))
                 params = [param_rows[0]]
             if on_conflict_suffix_sql:
+                if where_clause_to_sql:
+                    on_conflict_suffix_sql += f" WHERE {where_clause_to_sql}"
                 result.append(on_conflict_suffix_sql)
+                param_rows += [params_on_conflict]
```

Comment:
Pushing it down would make this more simple and, besides that, I see that this only covers the branch for the backends that support returning fields. Note the other cases further down in this method that should also be updated.

Reply:
i don't think i have to touch the other cases. i moved rename the compile_where_clause_to_sql method to _compile_where_clause_to_sql and move it to the basic operations, so each backend will inherit this method.